### PR TITLE
Split some generic utilities from the non-linear extension

### PR DESCRIPTION
**Pull request by ajreynol**
This is work towards correcting some issues with model generation in the non-linear solver.

This moves some of the utility functions out of the solver so that the code is more flexible to further refactor.

There are no significant code changes in this commit apart from added documentation.

### CODE REVIEW

File: src/theory/arith/arith_utilities.h [hunk 1]
```diff
@@ -302,6 +301,31 @@ inline Node mkPi()
   return NodeManager::currentNM()->mkNullaryOperator(
       NodeManager::currentNM()->realType(), kind::PI);
 }
+/** Join kinds, where k1 and k2 are arithmetic relations returns an
+ * arithmetic relation ret such that
+ * if (a <k1> b) and (a <k2> b), then (a <ret> b).
+ */
+Kind joinKinds(Kind k1, Kind k2);
```
Comment (ahmed-irfan) [inline src/theory/arith/arith_utilities.h]: What it is doing -- What is the contract of this function?

Comment (ahmed-irfan) [reply src/theory/arith/arith_utilities.h]: If I understood correctly from the code, Is it getting rid of redundant relation?

File: src/theory/arith/arith_utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  Assert(isRelationOperator(k2));
+  if (k1 == EQUAL)
+  {
+    return k2;
+  }
+  else if (k1 == LT)
+  {
+    if (k2 == LEQ)
+    {
+      return k1;
```
Comment (ahmed-irfan) [inline src/theory/arith/arith_utilities.cpp]: the behavior seems similar to joinKinds. I wonder if that is intentional

File: src/theory/arith/arith_utilities.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    }
+  }
+  else if (k1 == GT)
+  {
+    if (k2 == GEQ)
+    {
+      return k1;
```
Comment (ahmed-irfan) [inline src/theory/arith/arith_utilities.cpp]: ditto

File: src/theory/arith/arith_utilities.h (hunk 1 above)
Comment (ajreynol) [reply src/theory/arith/arith_utilities.h]: This is used in the lemma schema here: https://github.com/CVC4/CVC4/blob/master/src/theory/arith/nonlinear_extension.h#L828.

Basically, if you have two relations between related terms, you need to know what the resulting relation is.

File: src/theory/arith/arith_utilities.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/arith/arith_utilities.cpp]: The two are similar but trans applies to 3 terms whereas join applies to 2, for example `a = b ^ b >= c` imples` a >= c` whereas `a=b ^ a>=b` implies `a=b`.

Comment (ahmed-irfan) [reply src/theory/arith/arith_utilities.cpp]: I see

**Pull request merged**